File: Python/DP/coinChange.py
import sys
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        tmp = [sys.maxint] * (amount+1)
        tmp[0] = 0
        for i in xrange(amount+1):
            if tmp[i] != sys.maxint:
                for coin in coins:
                    if i+coin <= amount:
                        tmp[i+coin] = min(tmp[i+coin], tmp[i]+1)
        return tmp[amount] if tmp[amount] !=  sys.maxint else -1

File: src/DP/CoinChange.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package DP;

/**
 *
 * @author Sam
 */
public class CoinChange {
    public int coinChange(int[] coins, int amount) {
        int[] cache = new int[amount+1];
        cache[0] = 0;
        for( int i = 1; i <= amount; i++ ) cache[i] = Integer.MAX_VALUE;
        for( int i = 0; i <= amount; i++ ){
            for( int j = 0; j < coins.length; j++ ){
                if( i + coins[j] <= amount ){
                    cache[ i + coins[j] ] =  Math.min( cache[i + coins[j]], cache[i] + 1  ); 
                }
            }
        } 
        return cache[ amount ] == Integer.MAX_VALUE ? -1 : cache[amount];
    }
}
